### PR TITLE
Potential Flow Residual Loss: Bernoulli-consistency auxiliary signal

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

**PARADIGM-LEVEL CHANGE.** The model currently predicts raw CFD pressure at every node. But most of the pressure field is explained by classical inviscid aerodynamics — the Bernoulli equation: `p = p_inf + 0.5 * (U_inf^2 - |u|^2)`. The hard part is NOT the bulk pressure variation (well-described by potential flow) but the **nonlinear viscous correction**: separation, viscous displacement, wake interaction in tandem.

**Idea:** Add an auxiliary loss that penalizes the *Bernoulli residual* — the discrepancy between the model's predicted pressure and what Bernoulli says the pressure should be given the model's OWN predicted velocity. This couples velocity and pressure predictions through a known physical law, forcing thermodynamic consistency.

The model architecture is UNCHANGED — still predicts full (Ux, Uy, p). The Bernoulli residual is an **auxiliary loss signal** that focuses gradients on the nonlinear correction where errors are largest: LE suction peak, TE separation, tandem slot wake interaction.

**Why this is paradigm-level:** Currently the model predicts Ux, Uy, p independently — nothing enforces that these three fields are physically consistent. The Bernoulli residual loss creates a coupling: if the velocity prediction improves, the pressure target adjusts automatically. This is different from and complementary to the residual prediction (PR #1927, merged) which used a global baseline — this uses a **per-node, velocity-dependent** physics baseline.

**Key distinction from failed Bernoulli CONSISTENCY LOSS (PR #2224):** That added Bernoulli as a hard constraint (loss term penalizing |p + 0.5u^2 - const|), which is wrong in viscous/separated regions. This adds Bernoulli as an **auxiliary target reframing** — the model learns to predict Δp (the viscous correction), which is a smaller and smoother target. The model can still predict whatever pressure it wants; the auxiliary loss just provides gradient signal shaped by the physics.

**Literature:**
- Leer et al. "Residual learning for flow reconstruction" (ICLR 2024) — 20-35% improvement from residual-over-analytical-baseline
- Wu et al. "Physics-consistent DeepONet" (NeurIPS 2023) — Bernoulli consistency as auxiliary training signal

## Instructions

Add `--bernoulli_residual_loss` flag and `--bernoulli_weight` (float, default 0.1). After the forward pass, compute Bernoulli-based analytical pressure from the model's OWN velocity predictions, then add an auxiliary L1 loss on the residual.

### Step 1: Compute the Bernoulli analytical pressure

After the forward pass produces predictions for Ux, Uy, p:

```python
# IMPORTANT: Compute in the ORIGINAL physics space (before asinh normalization),
# then transform back to the normalized space for loss computation.

# Extract velocity predictions (denormalized / physics units)
u_pred = pred_denorm[:, :, 0]   # Ux in physics units [B, N]
v_pred = pred_denorm[:, :, 1]   # Uy in physics units [B, N]

# Freestream velocity magnitude (from input scalars — check how Umag is stored)
# Look for the Umag channel in the input features or global scalars
umag = batch_umag  # [B] or [B, 1] — freestream speed

# Bernoulli analytical pressure (relative to freestream):
# p_bernoulli = 0.5 * (U_inf^2 - (u^2 + v^2))
# (in non-dim Cp-like form, matching the normalization in _phys_norm)
p_bernoulli = 0.5 * (umag.unsqueeze(-1)**2 - u_pred**2 - v_pred**2)  # [B, N]

# Transform to the same space as the model's pressure prediction
# If using asinh normalization: p_bernoulli_norm = asinh(p_bernoulli / scale)
p_bernoulli_norm = torch.asinh(p_bernoulli / args.asinh_scale)  # match model's output space
```

### Step 2: Compute Bernoulli residual loss

```python
if args.bernoulli_residual_loss:
    # Detach the Bernoulli baseline so velocity gradients don't flow through it
    # (the auxiliary signal should shape pressure gradients, not velocity)
    p_bernoulli_detached = p_bernoulli_norm.detach()
    
    # Residual: how much the predicted pressure deviates from Bernoulli expectation
    delta_p_pred   = pred_p_norm - p_bernoulli_detached   # model's viscous correction
    delta_p_target = target_p_norm - p_bernoulli_detached  # true viscous correction
    
    bernoulli_loss = F.l1_loss(delta_p_pred, delta_p_target)
    
    # Add to total loss
    total_loss = total_loss + args.bernoulli_weight * bernoulli_loss
```

### Step 3: Handle PCGrad interaction

**CRITICAL:** From the lowrank experiment we learned that auxiliary losses added outside the PCGrad loss components (loss_A/B/C) only propagate ~20% of batches. To ensure the Bernoulli loss has full effect, add it INSIDE the pressure component of the PCGrad loss rather than as a separate term:

```python
# Inside the PCGrad loss computation, ADD bernoulli_loss to the pressure loss component
# e.g., if loss_A is the pressure loss:
if args.bernoulli_residual_loss:
    loss_A = loss_A + args.bernoulli_weight * bernoulli_loss
```

Check how `loss_A`, `loss_B`, `loss_C` are composed in the PCGrad section of `train.py` and add the Bernoulli term to whichever includes pressure loss.

### Step 4: Add flags

```python
parser.add_argument('--bernoulli_residual_loss', action='store_true',
                    help='Add Bernoulli residual auxiliary loss coupling velocity and pressure')
parser.add_argument('--bernoulli_weight', type=float, default=0.1,
                    help='Weight for Bernoulli residual loss (default 0.1)')
```

### Step 5: Log the Bernoulli residual metrics

```python
if args.bernoulli_residual_loss:
    wandb.log({
        'train/bernoulli_residual_loss': bernoulli_loss.item(),
        'train/bernoulli_mean_residual': delta_p_target.abs().mean().item(),
    })
```

### Step 6: Sanity checks before training

- Print `p_bernoulli_norm.mean(), p_bernoulli_norm.std()` on first batch — should be in similar range to model predictions
- Print `delta_p_target.abs().mean()` — this is the mean viscous correction. Should be smaller than `target_p_norm.abs().mean()` (the full pressure)
- If delta_p_target is LARGER than target_p, the normalization is wrong

### Training commands

```bash
# Seed 42
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent alphonse --seed 42 \
  --wandb_name "alphonse/potential-flow-residual-s42" \
  --wandb_group "potential-flow-residual" \
  --bernoulli_residual_loss --bernoulli_weight 0.1 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --re_extreme_weight 2.0

# Seed 73: same but CUDA_VISIBLE_DEVICES=1, --seed 73, --wandb_name "alphonse/potential-flow-residual-s73"
```

## Baseline

Current baseline (PR #2290, Re-Stratified Sampling, merged 2026-04-08, W&B-verified):

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| **p_in** | **11.742** | < 11.74 |
| p_oodc | 7.643 | < 7.64 |
| **p_tan** | **27.874** | < 27.87 |
| p_re | 6.419 | < 6.42 |

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### Experiment: Bernoulli Potential Flow Residual Loss (bernoulli_weight=0.1)

**Command used:**
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --wandb_name alphonse/potential-flow-residual-s42 --wandb_group potential-flow-residual --seed 42 --bernoulli_residual_loss --bernoulli_weight 0.1
cd cfd_tandemfoil && python train.py --agent alphonse --wandb_name alphonse/potential-flow-residual-s73 --wandb_group potential-flow-residual --seed 73 --bernoulli_residual_loss --bernoulli_weight 0.1
```

**W&B run IDs:** `lhksu0k1` (seed 42), `r25k1nd5` (seed 73)

**Peak memory:** ~0 MiB additional (no significant change from baseline)

### Metrics (2-seed average)

| Metric | Seed 42 | Seed 73 | **Mean** | Baseline (PR #2290) | Delta |
|--------|---------|---------|----------|---------------------|-------|
| val_in_dist mae_surf_p | 11.9 | 11.9 | **11.90** | 11.742 | +1.3% ↑ regression |
| val_tandem_transfer mae_surf_p | 28.2 | 28.8 | **28.50** | 27.874 | +2.2% ↑ regression |
| val_ood_cond mae_surf_p | 7.7 | 7.6 | **7.65** | 7.643 | +0.09% (flat) |
| val_ood_re mae_surf_p | 6.3 | 6.3 | **6.30** | 6.419 | **-1.9% ✓ improvement** |

Best epochs: 147 (seed 42), 145 (seed 73). Training time: ~180 min each.

### Sanity check (first batch)
`Cp_B mean (surface): 0.1146, mean |delta_p_target|: 0.4056` — Bernoulli baseline is reasonable, and the residual magnitude (~0.4) is non-trivial, confirming the viscous correction is a meaningful signal.

### What happened

The Bernoulli residual loss did not improve overall performance. p_in and p_tan both regressed modestly (~1-2%), while p_oodc was essentially flat and p_re improved by ~2%. This is a mixed signal: the p_re improvement suggests the viscous correction framing may have some utility for Re-extrapolation (where inviscid physics generalize better), but the in-distribution and tandem-transfer regressions indicate the auxiliary constraint is interfering with the main task.

**Likely causes of failure:**
1. **Weight may be too high**: bernoulli_weight=0.1 might be pulling gradients too aggressively away from direct L1 supervision. A smaller weight (0.01–0.05) could reduce interference.
2. **Bernoulli is a poor approximation at surface nodes**: The surface nodes include boundary layer effects and separation, where the inviscid Bernoulli equation breaks down most severely. The model is learning to correct this residual, but it's a noisy target and may conflict with the direct pressure supervision.
3. **Detached baseline correctness**: The Bernoulli baseline is computed from the *predicted* velocities (detached), not ground-truth velocities. This means the baseline quality depends on velocity prediction quality, which may introduce noise.

### Implementation note
The Bernoulli loss was integrated into PCGrad loss_a, loss_b, and _grp_loss (not just the main loss), ensuring full gradient propagation on all ~80% of batches where PCGrad does its own backward pass.

### Suggested follow-ups
1. **Lower bernoulli_weight to 0.01–0.05**: The 0.1 weight may be too aggressive; a softer regularizer could preserve the p_re benefit without regressing p_in/p_tan.
2. **Apply only to volume nodes, not surface**: Surface nodes violate Bernoulli assumptions most severely. Restricting to volume nodes might make the constraint more physically consistent.
3. **Use ground-truth velocities for Bernoulli baseline**: Compute Cp_B from y_norm (denormalized) rather than pred, making the residual a direct measure of viscous correction rather than a compounding velocity+pressure correction.
4. **Curriculum scheduling**: Start bernoulli_weight at 0.0 and ramp up after epoch 50, letting the model establish good velocity predictions before enforcing the physics constraint.

**Comment by tcapelle** [comment]
ADVISOR: Promising direction — the **p_re improvement of -1.9% (6.30 vs 6.419)** is the best p_re result since PR #2213 (6.300). The physics coupling between velocity and pressure IS providing useful regularization for Re-extrapolation. But the weight is too aggressive, causing p_in/p_tan regressions.

**Please rerun with these 3 changes:**

1. **Lower weight to 0.03**: `--bernoulli_weight 0.03` (vs 0.1 before). This should preserve the p_re benefit while reducing interference on p_in/p_tan.

2. **Apply ONLY to volume nodes**: Exclude surface nodes from the Bernoulli residual loss. Surface nodes are in the boundary layer where Bernoulli breaks down most severely — including them adds noise. Add a `~surface_mask` filter when computing the Bernoulli residual.

3. **Use GROUND-TRUTH velocities for Bernoulli baseline**: Compute `p_bernoulli = 0.5 * (umag^2 - u_true^2 - v_true^2)` from the target velocity (denormalized), not the predicted velocity. This eliminates noise from velocity prediction quality and makes the residual a clean measure of viscous correction.

Use the same wandb_group (`potential-flow-residual`) so we can compare runs. Run 2 seeds as before.

```bash
--bernoulli_residual_loss --bernoulli_weight 0.03
```

Keep everything else the same. This is the most promising paradigm-level direction we've found — let's optimize it.

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results (v2 — Revised per advisor feedback)

### Changes from v1:
1. **bernoulli_weight reduced** from 0.1 → 0.03
2. **Volume nodes only** (excluded surface nodes where Bernoulli breaks down)
3. **Ground-truth velocities** for Bernoulli baseline (y_norm instead of pred)

### Metrics (2-seed average)

| Metric | Seed 42 | Seed 73 | **Mean** | Baseline (PR #2290) | Delta |
|--------|---------|---------|----------|---------------------|-------|
| val_in_dist mae_surf_p | 11.9 | 12.0 | **11.95** | 11.742 | +1.8% ↑ regression |
| val_tandem_transfer mae_surf_p | 28.4 | 28.5 | **28.45** | 27.874 | +2.1% ↑ regression |
| val_ood_cond mae_surf_p | 7.8 | 7.4 | **7.60** | 7.643 | -0.6% ✓ improvement |
| val_ood_re mae_surf_p | 6.5 | 6.5 | **6.50** | 6.419 | +1.3% ↑ regression |

**W&B run IDs:** `aguyyx0e` (seed 42), `go4ly0kq` (seed 73)
**W&B group:** `potential-flow-residual`
**Best epochs:** ~148 (s42), ~147 (s73). Training time: ~180 min each.

### Sanity checks
`[Bernoulli sanity] Cp_B mean (volume): 0.0445, mean |delta_p_target|: 0.4800, bernoulli_loss: 0.3192` (seed 42)
`[Bernoulli sanity] Cp_B mean (volume): 0.2044, mean |delta_p_target|: 0.4334, bernoulli_loss: 0.2932` (seed 73)

### Analysis

The v2 revision **did not improve** over v1 or baseline. Compared to v1 (bernoulli_weight=0.1, surface nodes, pred velocities):

| Metric | v1 avg | v2 avg | Direction |
|--------|--------|--------|-----------|
| p_in | 11.90 | 11.95 | Slightly worse |
| p_tan | 28.50 | 28.45 | Slightly better |
| p_oodc | 7.65 | 7.60 | Slightly better |
| p_re | **6.30** | 6.50 | **Worse** ← lost the best result from v1 |

The most notable finding is that **p_re regressed from 6.30 (v1) to 6.50 (v2)**. The v1 configuration with predicted velocities on surface nodes actually worked better for Re-extrapolation. This suggests:

1. **Using ground-truth velocities removed a useful coupling signal**: When Bernoulli is computed from predicted velocities, errors in velocity directly shape the pressure residual — this created an implicit velocity-pressure consistency gradient that aided OOD-Re generalization.
2. **Volume nodes are less informative than surface nodes for pressure physics**: The viscous correction is largest at the surface (boundary layer, separation). Volume nodes are closer to inviscid flow where the residual is small and less informative.
3. **The lower weight (0.03 vs 0.1) may have been insufficient** to overcome the reduced signal from volume-only application.

### Conclusion

The Bernoulli residual loss approach overall shows mixed results:
- v1 achieved a genuine p_re improvement (-1.9%) but regressed p_in/p_tan
- v2 lost the p_re benefit and still regressed p_in/p_tan
- Neither version improved the key metrics (p_in, p_tan) needed for a merge

This suggests the Bernoulli residual framing provides some useful physics-based regularization for Re-extrapolation (when applied to surface with predicted velocities), but the current formulation interferes with the main loss optimization. The approach may be fundamentally limited because the inviscid assumption breaks down most where it matters most.

### Suggested follow-ups (if pursuing further)
1. **Hybrid approach**: v1 settings (pred velocities, surface, weight=0.1) but with curriculum — start at weight=0 and ramp up after epoch 50
2. **Bernoulli consistency loss on velocity only**: Instead of pressure residual, penalize |Ux_pred^2 + Uy_pred^2 - (Umag^2 - 2*Cp_pred)| — coupling velocity and pressure in the other direction
3. **Close this direction**: The modest p_re improvement doesn't compensate for p_in/p_tan regressions. GPU time may be better spent on other approaches.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2299 because ADVISOR: Closing after 2 iterations. Neither v1 nor v2 beat baseline on p_in/p_tan, which are our merge-critical metrics.

**Key finding from this experiment:** The velocity-pressure coupling from v1 (predicted velocities on surface, weight=0.1) produced a genuine p_re=-1.9% improvement (our best p_re since PR #2213). But the same mechanism interferes with the main L1 supervision on p_in/p_tan. v2's changes (GT velocities, volume-only, lower weight) lost the p_re benefit by removing the coupling gradient.

**Insight for future work:** The Bernoulli framing CAN help Re-extrapolation through implicit velocity-pressure consistency — but the physics constraint needs to be applied as a soft regularizer with curriculum scheduling (not a fixed-weight loss) to avoid interfering with the main task. This is a known challenge in physics-informed ML and warrants revisiting only if a principled curriculum schedule is designed.

Two iterations is sufficient to establish the limits of this approach. Well-executed experiment — the v1 p_re result is a useful data point. Moving you to a fresh direction..
